src/main/scala/firrtl/AddDescriptionNodes.scala

```diff
@@ -71,44 +125,85 @@ class AddDescriptionNodes extends Transform {
+      case p @ Port(_, name, _, _) if compMap.contains(name) =>
+        name -> compMap(name)
+    }.toMap
 
-    val modDesc = modMap.get(newMod.name).map {
-      desc => DocString(StringLit.unescape(desc.mkString("\n\n")))
-    }
+    val modDesc = modMap.get(newMod.name).getOrElse(Seq())
 
     if (portDesc.nonEmpty || modDesc.nonEmpty) {
-      DescribedMod(modDesc.getOrElse(EmptyDescription), portDesc, newMod)
+      DescribedMod(modDesc, portDesc, newMod)
     } else {
       newMod
     }
   }
 
-  def collectMaps(annos: Seq[Annotation]): (Map[String, Seq[String]], Map[String, Map[String, Seq[String]]]) = {
-    val modMap = annos.collect {
-      case DescriptionAnnotation(ModuleName(m, CircuitName(c)), desc) => (m, desc)
-    }.groupBy(_._1).mapValues(_.map(_._2))
+  /**
+    * Merges descriptions of like types.
+    *
+    * Multiple DocStrings on the same object get merged together into one big multi-line comment.
+    * Similarly, multiple attributes on the same object get merged into one attribute with attributes separated by
+    * commas.
+    * @param descs List of `Description`s that are modifying the same object
+    * @return List of `Description`s with some descriptions merged
+    */
+  def mergeDescriptions(descs: Seq[Description]): Seq[Description] = {
+    val (docs: Seq[DocString], nodocs) = descs.partition(_ match {
+      case _: DocString => true
+      case _ => false
+    })
+    val (attrs: Seq[Attribute], rest) = nodocs.partition(_ match {
+      case _: Attribute => true
+      case _ => false
+    })
+
+    val doc = if (docs.nonEmpty) {
+      Seq(DocString(StringLit.unescape(docs.map(_.string.string).mkString("\n\n"))))
+    } else {
+      Seq()
+    }
+    val attr = if (attrs.nonEmpty) {
+      Seq(Attribute(StringLit.unescape(attrs.map(_.string.string).mkString(", "))))
+    } else {
+      Seq()
+    }
+
+    rest ++ doc ++ attr
+  }
+
+  def collectMaps(annos: Seq[Annotation]): (Map[String, Seq[Description]], Map[String, Map[String, Seq[Description]]]) = {
+    val modMap: Map[String, Seq[Description]] = annos.collect {
+      case DocStringAnnotation(ModuleName(m, _), desc) => (m, DocString(StringLit.unescape(desc)))
+      case AttributeAnnotation(ModuleName(m, _), desc) => (m, Attribute(StringLit.unescape(desc)))
+    }.groupBy(_._1).mapValues(_.map(_._2)).mapValues(mergeDescriptions)
 
     val compMap = annos.collect {
-      case DescriptionAnnotation(ComponentName(comp, ModuleName(mod, CircuitName(circ))), desc) =>
-        (mod, comp, desc)
-    }.groupBy(_._1).mapValues(_.groupBy(_._2).mapValues(_.map(_._3)))
+      case DocStringAnnotation(ComponentName(c, ModuleName(m, _)), desc) =>
+        (m, c, DocString(StringLit.unescape(desc)))
+      case AttributeAnnotation(ComponentName(c, ModuleName(m, _)), desc) =>
+        (m, c, Attribute(StringLit.unescape(desc)))
+    }.groupBy(_._1).mapValues(_.groupBy(_._2).mapValues(_.map(_._3)).mapValues(mergeDescriptions))
```

Comment:
👀 Is this something we could make more clear? Maybe just in a function?

Reply:
Scala 2.13 has `groupMap`, which seems like it's defined to deal with some of this.

I broke it down a bit and added some comments, which hopefully makes it a bit easier to understand.